core: Better communicate errors to the CLI

IagoAbal opened the pull request:
- Add backtrace to fatal errors to help diagnosing bug reports (Pfff).
- Report rule evaluation errors instead of "silently" logging them.

Helps #3547

test plan:

    % cat fatal.py
    def foo():
    return 1
    }
    % semgrep -l py -e 'x' fatal.py
    running 1 rules...
    semgrep-core reported a fatal error:
    -----
    Fatal Error: (Failure "Lexer_python.top_mode: empty stack")
    Raised at Parse_target.run in file "src/parsing/Parse_target.ml", line 176, characters 17-26
    ...
    -----
    Please file a bug report at https://github.com/returntocorp/semgrep/issues/new/choose
    ...

    % cat warn.yaml
    rules:
    - id: test
      languages: [python]
      patterns:
      - pattern: $X
      - metavariable-pattern:
          metavariable: $Y
          pattern: x
      message: Test
      severity: ERROR
    % cat warn.py
    x
    % semgrep -c warn.yaml -v warn.py
    ...
    semgrep-core reported a matching error
    --> matching internal error: rule test: metavariable-pattern failed because $Y it not in scope, please check your rule
    ...



PR checklist:
- [x] documentation is up to date
- [x] changelog is up to date

IagoAbal commented:
@brendongo you or someone who is more involved in the CLI should check that you like my changes to `core_exception.py` and `errors.py`.

emjin commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -111,15 +112,29 @@ type env = {
+ * as we did before. *)
+let error env msg =
+  (* We are not supposed to report errors in the config file for several reasons
+   * (one being that it's often a temporary file anyways, so we report them on
```
I think you want an end parenthesis after the "anyways"?

emjin commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -111,15 +112,30 @@ type env = {
+   * the target file. *)
+  let loc = PI.first_loc_of_file env.file in
+  let s = Printf.sprintf "rule %s: %s" env.rule_id msg in
+  (* TODO: warning or error? MatchingError or ... ? *)
```
A little confused---it sounds like these used to be silently logged. Will reporting them as errors now cause them to crash Semgrep? Or is this just a cosmetic question?

IagoAbal replied:
No it won't crash Semgrep, they will just be communicated to the user instead of "silently" logged. Now that I'm looking into the test failures, I'm wondering if maybe not all of them should be communicated though!

IagoAbal replied:
When running Semgrep you will e.g. see:
```
found problems analyzing 1 files; run with --verbose for details or run with --strict to exit non-zero if any file cannot be analyzed cleanly
```
And using `--verbose` you'll get:
```
semgrep-core reported a matching error
  --> matching internal error: rule test: metavariable-pattern failed because $Y it not in scope, please check your rule
```
But any matches that were found and reported previously should continue to be found and reported now.

emjin commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -644,19 +660,20 @@ and satisfies_metavar_pattern_condition env r mvar opt_xlang formula =
   (* If anything goes wrong the default is to filter out! *)
   match List.assoc_opt mvar bindings with
   | None ->
-      (* TODO: Report a warning to the user? *)
```
Yay!

emjin commented on semgrep/semgrep/core_exception.py:
```diff
@@ -112,12 +114,13 @@ def into_semgrep_error(self) -> SemgrepError:
-                    if self._check_id != "ParseError" and "message" in self._extra
-                    else ""
-                ),
+                long_msg=f"Could not parse {self._path.name} as {self._language}",
```
Is `self._extra['message']` ever informative with ParseErrors? Maybe worth leaving it in

IagoAbal replied:
Note that before there was a `self._check_id != "ParseError"`. In the few tests I did I just got a "Syntax error" message which is not very useful.

aryx commented on semgrep/tests/e2e/snapshots/test_check/test_timeout_threshold/error.txt:
```diff
@@ -7,4 +7,4 @@ rules:
 Semgrep stopped running rules on targets/equivalence/open_redirect.py after 1 timeout error(s). See `--timeout-threshold` for more info.
 ran 3 rules on 1 files: 0 findings
-1 files could not be analyzed; run with --verbose for details or run with --strict to exit non-zero if any file cannot be analyzed
+found problems analyzing 1 files; run with --verbose for details or run with --strict to exit non-zero if any file cannot be analyzed cleanly
```
we could fix the plural, it's 1 file, not 1 files :)

aryx commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -111,15 +112,30 @@ type env = {
   xlang : R.xlang;
   equivalences : Equivalence.equivalences;
+  (* problems found during evaluation, one day these may be caught earlier by
+   * the meta-checker *)
```
yep, exactly, for each of those errors now surfaced to the user we should improve the metachecker to detect them too
statically.

aryx commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -739,20 +758,21 @@ and satisfies_metavar_pattern_condition env r mvar opt_xlang formula =
-                "rule %s: metavariable-pattern: the content of %s is not text"
-                env.rule_id mvar;
+          | Some _lang, mval ->
+              logger#debug
```
why not an error for this one?

IagoAbal replied:
Initially I thought it should be an error too, but it can happen easily without being an error, I saw this when running the changes on semgrep-rules. I added this comment:
```ocaml
(* This is not necessarily an error in the rule, e.g. you may be
* matching `$STRING + ...` and then add a metavariable-pattern on
* `$STRING`. This will only work when `$STRING` binds to some text
* but it can naturally bind to other string expressions. *)
```

brendongo commented on semgrep/semgrep/output.py:
```diff
@@ -307,9 +307,9 @@ def close(self) -> None:
+            num_errors = len(self.semgrep_structured_errors)
+            plural = "s" if num_errors > 1 else ""
+            error_stats = f"found problems analyzing {num_errors} file{plural}"
             final_error = self.semgrep_structured_errors[-1]
```
I think with this hacky assumption to have final_error be the last element of semgrep_structured_errors and CoreWarning having a 0 return code and https://github.com/returntocorp/semgrep/blob/develop/semgrep/semgrep/__main__.py#L25 I think there might be a scenario when semgrep should be exiting with non-zero exit code but instead has a 0 exit code?

IagoAbal replied:
Good catch, I don't have an example of that but it seems possible to me. I'll give it a try at fixing it.

IagoAbal committed:
core: Better communicate errors to the CLI

- Add backtrace to fatal errors to help diagnosing bug reports (Pfff).
- Report rule evaluation errors instead of "silently" logging them.

Helps #3547

test plan:

    % cat fatal.py
    def foo():
    return 1
    }
    % semgrep -l py -e 'x' fatal.py
    running 1 rules...
    semgrep-core reported a fatal error:
    -----
    Fatal Error: (Failure "Lexer_python.top_mode: empty stack")
    Raised at Parse_target.run in file "src/parsing/Parse_target.ml", line 176, characters 17-26
    ...
    -----
    Please file a bug report at https://github.com/returntocorp/semgrep/issues/new/choose
    ...

    % cat warn.yaml
    rules:
    - id: test
      languages: [python]
      patterns:
      - pattern: $X
      - metavariable-pattern:
          metavariable: $Y
          pattern: x
      message: Test
      severity: ERROR
    % cat warn.py
    x
    % semgrep -c warn.yaml -v warn.py
    ...
    semgrep-core reported a matching error
    --> matching internal error: rule test: metavariable-pattern failed because $Y it not in scope, please check your rule
    ...

IagoAbal commented on semgrep/semgrep/error.py:
```diff
@@ -342,6 +343,47 @@ def to_dict_base(self) -> Dict[str, Any]:
+    check_id: str = attr.ib()
+    msg: str = attr.ib()
+
+    code = INVALID_PATTERN_EXIT_CODE
```
@brendongo Despite this Semgrep still returns 0 when one of these warnings is reported, is that normal?

IagoAbal replied:
I'll merge but let me know if I should do something about this.
